### PR TITLE
Refactoring Cassandra visibility persistence manager for NoSQL support

### PR DESCRIPTION
**Pull request by longquanzheng**
**What changed?**
1. Create NoSQL interface for visibility  operation
2. Extract the logic of Cassandra to implement the interface
3. Replace visibiltyStore with NoSQL plugin db style

**Why?**
For https://github.com/uber/cadence/issues/3514

**How did you test it?**
existing tests

**Potential risks**
Low risks

**Release notes**
No

**Documentation Changes**
No

### CODE REVIEW

File: common/persistence/cassandra/cassandraVisibilityPersistence.go [hunk 1]
```diff
@@ -25,165 +25,26 @@ import (
...
-		`WHERE domain_id = ? ` +
-		`AND domain_partition = ? ` +
-		`AND close_time >= ? ` +
-		`AND close_time <= ? ` +
-		`AND status = ? `
-)
-
 type (
-	cassandraVisibilityPersistence struct {
+	nosqlVisibilityManager struct {
```
Comment (yux0) [inline common/persistence/cassandra/cassandraVisibilityPersistence.go:45]: maybe also update the file name

**Review by yux0 — APPROVED** [review]

**Review by yycptt — APPROVED** [review]
LGTM overall. Just some nit comments

File: common/persistence/cassandra/cassandraVisibilityPersistence.go (hunk 1 above)
Comment (yycptt) [reply common/persistence/cassandra/cassandraVisibilityPersistence.go:45]: I guess that will be done when we move them to the nosql package? Currently they are hardcoded to use cassandra plugin.

Comment (longquanzheng) [reply common/persistence/cassandra/cassandraVisibilityPersistence.go:45]: Yes that's the plan.

File: common/persistence/cassandra/cassandraVisibilityPersistence.go [hunk 2]
```diff
@@ -25,165 +25,26 @@ import (
...
-	"github.com/uber/cadence/common/types/mapper/thrift"
 )
 
 // Fixed domain values for now
 const (
-	domainPartition        = 0
 	defaultCloseTTLSeconds = 86400
 	openExecutionTTLBuffer = int64(86400) // setting it to a day to account for shard going down
 
 	maxCassandraTTL = int64(157680000) // Cassandra max support time is 2038-01-19T03:14:06+00:00. Updated this to 5 years to support until year 2033
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/cassandra/cassandraVisibilityPersistence.go:41]: It's also used by taskStore, I will remove it in https://github.com/uber/cadence/pull/4229 later.

File: common/persistence/nosql/nosqlplugin/interfaces.go [hunk 3]
```diff
@@ -203,6 +204,78 @@ type (
...
+		DomainID string
+		// NOTE: this is only for some implementation (e.g. Cassandra) that uses multiple tables,
+		// they needs to delete record from the open execution table. Ignore this field if not need it
+		UpdateOpenToClose bool
+		//  Similar as UpdateOpenToClose
+		UpdateCloseToOpen bool
+	}
+
+	// TODO separate in the future when need it
+	VisibilityRowForRead = persistence.InternalVisibilityWorkflowExecutionInfo
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/nosql/nosqlplugin/interfaces.go:258]: yeah, that fits better.

File: common/persistence/nosql/nosqlplugin/cassandra/constants.go [hunk 4]
```diff
@@ -32,6 +34,14 @@ var (
 	defaultVisibilityTimestamp = p.UnixNanoToDBTimestamp(defaultDateTime.UnixNano())
 )
 
+const (
+	// Cassandra max support time is 2038-01-19T03:14:06+00:00. Updated this to 5 years to support until year 2033
+	// See https://github.com/uber/cadence/issues/4200
+	maxCassandraTTL = int64(157680000)
+
+	cassandraLowConslevel = gocql.One
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/nosql/nosqlplugin/cassandra/constants.go:42]: yeah good idea.

**Pull request merged**